src/shims/windows/fs.rs

```diff
@@ -390,6 +391,253 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
+        let buf = this.read_pointer(buf)?;
+        let count = this.read_scalar(n)?.to_u32()? as usize;
+        let byte_offset = this.read_target_usize(byte_offset)?; // is actually a pointer
+        let key = this.read_pointer(key)?;
+        let io_status_block =
+            this.deref_pointer_as(io_status_block, this.windows_ty_layout("IO_STATUS_BLOCK"))?;
+
+        if event != Handle::Null {
+            throw_unsup_format!("`NtReadFile` `Event` parameter is non-null, which is unsupported");
+        }
+
+        if !this.ptr_is_null(apc_routine)? {
+            throw_unsup_format!(
+                "`NtReadFile` `ApcRoutine` parameter is not null, which is unsupported"
+            );
+        }
+
+        if !this.ptr_is_null(apc_ctx)? {
+            throw_unsup_format!(
+                "`NtReadFile` `ApcContext` parameter is not null, which is unsupported"
+            );
+        }
+
+        if byte_offset != 0 {
+            throw_unsup_format!(
+                "`NtReadFile` `ByteOffset` parameter is non-null, which is unsupported"
+            );
+        }
+
+        if !this.ptr_is_null(key)? {
+            throw_unsup_format!("`NtReadFile` `Key` parameter is not null, which is unsupported");
+        }
+
+        let io_status = {
+            let anon = this.project_field_named(&io_status_block, "Anonymous")?;
+            this.project_field_named(&anon, "Status")?
+        };
+        let io_status_info = this.project_field_named(&io_status_block, "Information")?;
+
+        let finish = {
+            let io_status = io_status.clone();
+            let io_status_info = io_status_info.clone();
+            callback!(
+                @capture<'tcx> {
+                    count: usize,
+                    io_status: MPlaceTy<'tcx>,
+                    io_status_info: MPlaceTy<'tcx>,
+                }
+                |this, result: Result<usize, IoError>| {
+                    match result {
+                        Ok(read_size) => {
+                            assert!(read_size <= count);
+                            // This must fit since `count` fits.
+                            this.write_int(u64::try_from(read_size).unwrap(), &io_status_info)?;
+                            this.write_int(0, &io_status)
+                        }
+                        Err(e) => {
+                            this.write_int(0, &io_status_info)?;
+                            this.write_int(error_kind_to_ntstatus(e), &io_status)
+                        }
+                }}
+            )
+        };
+
+        let fd = match handle {
+            Handle::File(fd) => fd,
+            _ => this.invalid_handle("NtWriteFile")?,
+        };
+
+        let Some(desc) = this.machine.fds.get(fd) else { this.invalid_handle("NtReadFile")? };
+
+        // TODO: Windows expects this function to return its error, not set the global one
+        desc.read(this.machine.communicate(), buf, count, this, finish)?;
```

Comment:
Does this comment only apply to `read`, not `write`?

Also, what to do with the error is entirely up to you -- the callback receives the error info. Your callback doesn't seem to set the global error so I'm not sure what the comment refers to?

Seems worth having a test which checks that this does indeed not clobber the global error.

Reply:
Accidentally copied over from an older version of the code. I'll add a test for that behavior.